File: examples/centroiding_models/run_centroiding.cpp
/**
 * @file run_centroiding.cpp
 * @brief Run independent image frames through a centroiding model with optional JSON output.
 *
 * The executable owns command-line collection, OpenCV image decoding, and
 * stable textual output. Model execution remains behind `CModelFacade`, while
 * example-side helpers own the selected model's grayscale and coordinate
 * semantics.
 */

#include "centroiding_io.h"
#include <chrono>
#include <memory>

#include <inference/inference_config_parsing.h>
#include <inference/model_facade.h>
#include <utils/logging/CLogger.h>

#include <opencv2/imgcodecs.hpp>
#include <tclap/CmdLine.h>

#include <exception>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace
{
    namespace demo = ptafdeploy::examples::centroiding_models;
    namespace fs = std::filesystem;
    namespace infer = ptafdeploy::inference;
    namespace logging = ptafdeploy::logging;

    /** @brief Return the process-wide logger configured from the environment once. */
    [[nodiscard]] logging::CLogger& GetLogger()
    {
        static logging::CLogger logger("run_centroiding", logging::ELogLevel::Info,
                                       logging::ELogColorMode::Disabled, std::clog, std::clog);
        static const bool environment_applied = logger.setLevelFromEnvironment();
        static_cast<void>(environment_applied);
        return logger;
    }

    /** @brief Validated command-line state for one image or sequence invocation. */
    struct SArguments
    {
        /** @brief Centroiding `.ptafmodel` manifest or raw `.onnx` artifact. */
        fs::path model_path{};

        /** @brief One image or directory, decoded frame by frame into uint8 grayscale. */
        fs::path image_path{};
        /** @brief Optional new or empty output directory. */
        fs::path output_path{};
        /** @brief Whether annotated images accompany the JSON report. */
        bool overlays{false};

        /** @brief Backend-neutral runtime used for raw models or explicit overrides. */
        infer::SRuntimeConfig runtime{};

        /** @brief Whether explicit CLI runtime values replace manifest policy. */
        bool runtime_overridden{false};
    };

    /**
     * @brief Declare, collect, and validate one centroiding command line.
     * @param argc Number of command-line arguments.
     * @param argv Command-line argument values owned by the process.
     * @return Validated invocation state independent of TCLAP types.
     * @throws TCLAP::ExitException When help or version output requests an early exit.
     * @throws TCLAP::ArgException If TCLAP rejects an option or value.
     * @throws std::invalid_argument If a runtime value violates the public contract.
     */
    [[nodiscard]] SArguments ParseArguments(const int argc, char** argv)
    {
        TCLAP::CmdLine command(
            "Run an image or directory through an image-only centroiding model. The model may be a "
            "role manifest or a raw ONNX artifact; image decoding, grayscale conversion, "
            "resize, and unit scaling are handled by this integration.",
            ' ', PTAFDEPLOY_CLI_VERSION);
        command.setExceptionHandling(false);

        // TCLAP remains private to argument collection. Runtime target parsing
        // uses the same backend-neutral API as manifests and other programs.
        TCLAP::UnlabeledValueArg<std::string> model_path(
            "model", "Path to a centroiding manifest or raw ONNX artifact", true, "",
            "model.ptafmodel|model.onnx", command);
        TCLAP::UnlabeledValueArg<std::string> image_path(
            "input_path", "Path to one image or a directory of images", true, "", "image.png",
            command);
        TCLAP::ValueArg<std::string> targets("", "targets",
                                             "Comma-separated execution-target priority override",
                                             false, "", "cpu,cuda,tensorrt", command);
        TCLAP::ValueArg<int> device("", "device", "Non-negative runtime device index override",
                                    false, 0, "non-negative integer", command);
        TCLAP::ValueArg<int> intra_op_threads(
            "", "intra-op-threads", "Runtime intra-operation threads; zero uses its default", false,
            1, "non-negative integer", command);
        TCLAP::ValueArg<int> inter_op_threads(
            "", "inter-op-threads", "Runtime inter-operation threads; zero uses its default", false,
            1, "non-negative integer", command);
        TCLAP::SwitchArg no_fallback("", "no-fallback",
                                     "Reject fallback below the requested execution targets",
                                     command, false);

        TCLAP::ValueArg<std::string> output_path("", "output", "New or empty results directory",
                                                 false, "", "PATH", command);
        TCLAP::SwitchArg overlays("", "overlays", "Save original-resolution crosshairs", command,
                                  false);
        command.parse(argc, argv);
        if (overlays.getValue() && output_path.getValue().empty())
            throw std::invalid_argument("--overlays requires --output");

        SArguments arguments;
        arguments.model_path = model_path.getValue();
        arguments.image_path = image_path.getValue();
        arguments.output_path = output_path.getValue();
        arguments.overlays = overlays.getValue();
        arguments.runtime.SetDeviceId(device.getValue());
        arguments.runtime.SetThreadCounts(intra_op_threads.getValue(), inter_op_threads.getValue());
        arguments.runtime.SetAllowFallback(!no_fallback.getValue());
        arguments.runtime.SetLogId("run_centroiding");
        if (targets.isSet())
        {
            arguments.runtime.execution_target_priority =
                infer::ParseExecutionTargetPriority(targets.getValue());
        }

        arguments.runtime_overridden = targets.isSet() || device.isSet() ||
                                       intra_op_threads.isSet() || inter_op_threads.isSet() ||
                                       no_fallback.getValue();
        return arguments;
    }

    /**
     * @brief Load a manifest or raw ONNX through the role-level facade.
     * @param arguments Validated model path and runtime policy.
     * @param model Facade populated only if model loading succeeds.
     * @throws std::exception If the artifact, manifest, or runtime is invalid.
     */
    void LoadModel(const SArguments& arguments, infer::CModelFacade& model)
    {
        const std::string extension = arguments.model_path.extension().string();
        if (extension == ".ptafmodel")
        {
            if (arguments.runtime_overridden)
            {
                model.LoadModelConfigWithRuntimeConfig(arguments.model_path.string(),
                                                       arguments.runtime);
            }
            else
            {
                model.LoadModelConfig(arguments.model_path.string());
            }
            return;
        }
        if (extension == ".onnx")
        {
            model.LoadModelWithRoleAndRuntimeConfig(
                arguments.model_path.string(), infer::EModelRole::centroiding, arguments.runtime);
            return;
        }

        throw std::invalid_argument("Centroiding expects a .ptafmodel manifest or .onnx artifact.");
    }

    /**
     * @brief Format tensor dimensions as a stable bracketed list.
     * @param shape Dimensions in logical tensor order.
     * @return Text such as `[1,1,1536,2048]`.
     */
    [[nodiscard]] std::string FormatShape(const std::vector<int64_t>& shape)
    {
        std::ostringstream stream;
        stream << '[';
        for (size_t index = 0U; index < shape.size(); ++index)
        {
            if (index != 0U)
            {
                stream << ',';
            }
            stream << shape[index];
        }
        stream << ']';
        return stream.str();
    }

    /**
     * @brief Print stable model, image, and centroid result fields to stdout.
     * @param model Loaded centroiding facade.
     * @param input Prepared model input tensor.
     * @param output Raw normalized centroid output.
     * @param image_size Original decoded image extent.
     * @param result Validated centroid in all supported coordinate spaces.
     */
    void PrintResult(const infer::CModelFacade& model, const infer::SFloatTensor& input,
                     const infer::SFloatTensor& output, const cv::Size image_size,
                     const demo::SCentroidResult& result)
    {
        std::cout << std::setprecision(9) << "role=" << model.GetRole() << '\n'
                  << "backend=" << model.GetBackendDetail() << '\n'
                  << "input_name=" << input.name << '\n'
                  << "input_shape=" << FormatShape(input.shape) << '\n'
                  << "image_size=[" << image_size.width << ',' << image_size.height << "]\n"
                  << "output_name=" << output.name << '\n'
                  << "output_shape=" << FormatShape(output.shape) << '\n'
                  << "centroid.normalized_x=" << result.normalized.x << '\n'
                  << "centroid.normalized_y=" << result.normalized.y << '\n'
                  << "centroid.model_input_x_px=" << result.model_input_pixels.x << '\n'
                  << "centroid.model_input_y_px=" << result.model_input_pixels.y << '\n'
                  << "centroid.original_x_px=" << result.original_image_pixels.x << '\n'
                  << "centroid.original_y_px=" << result.original_image_pixels.y << '\n';
    }
} // namespace

/**
 * @brief Run independent frames through a facade-backed centroiding model.
 * @param argc Number of command-line arguments.
 * @param argv Command-line argument values.
 * @return Zero on success, one for invalid user input, or two for runtime failures.
 */
int main(const int argc, char** argv)
{
    try
    {
        const SArguments arguments = ParseArguments(argc, argv);
        const auto frames = demo::SelectFrames(arguments.image_path);
        std::unique_ptr<demo::CReport> report;
        if (!arguments.output_path.empty())
        {
            demo::PrepareOutput(arguments.output_path, arguments.image_path);
            report = std::make_unique<demo::CReport>(
                arguments.output_path, demo::MetadataJson(arguments.image_path, frames.size(),
                                                          nullptr, arguments.model_path));
        }
        size_t index = frames.size();
        std::string stage = "model_load";
        try
        {
            infer::CModelFacade model;
            LoadModel(arguments, model);
            if (model.GetRole() != "centroiding" || model.GetNumInputs() != 1U ||
                model.GetNumOutputs() != 1U)
                throw std::invalid_argument("Centroiding requires one image input and one output");
            if (report)
                report->metadata = demo::MetadataJson(arguments.image_path, frames.size(), &model,
                                                      arguments.model_path);
            if (arguments.overlays)
                fs::create_directory(arguments.output_path / "overlays");
            for (index = 0; index < frames.size(); ++index)
            {
                stage = "decode";
                const auto& source = frames[index];
                const cv::Mat image = cv::imread(source.string(), cv::IMREAD_GRAYSCALE);
                if (image.empty())
                    throw std::runtime_error("Cannot decode image: " + source.string());
                stage = "preprocessing";
                const auto input = demo::PrepareImageTensor(image, model.GetInputInfo(0));
                stage = "inference";
                const auto start = std::chrono::steady_clock::now();
                const auto output = model.InferSingleFloatTensor(input);
                const double duration = std::chrono::duration<double, std::milli>(
                                            std::chrono::steady_clock::now() - start)
                                            .count();
                stage = "decoding_output";
                const auto result = demo::DecodeCentroid(
                    output,
                    cv::Size{static_cast<int>(input.shape[3]), static_cast<int>(input.shape[2])},
                    image.size());
                std::cout << "frame.index=" << index
                          << "\nframe.source=" << source.filename().string()
                          << "\ninference_ms=" << duration << '\n';
                PrintResult(model, input, output, image.size(), result);
                std::string overlay;
                if (arguments.overlays)
                {
                    stage = "overlay";
                    std::ostringstream name;
                    name << "overlays/" << std::setfill('0') << std::setw(6) << index << '_'
                         << source.stem().string() << ".png";
                    overlay = name.str();
                    demo::SaveOverlay(source, arguments.output_path / overlay, result,
                                      image.size());
                }
                stage = "report";
                if (report)
                    report->Append(demo::FrameJson(index, source.filename(), image.size(), output,
                                                   result, duration, overlay));
            }
            if (report)
                report->Publish(true);
        }
        catch (const std::exception& error)
        {
            if (report)
            {
                rapidjson::StringBuffer buffer;
                demo::JsonWriter w(buffer);
                w.StartObject();
                demo::String(w, "stage", stage);
                const bool frame_failed = stage != "model_load" && index < frames.size();
                w.Key("frame_index");
                if (frame_failed)
                    w.Uint64(index);
                else
                    w.Null();
                w.Key("source");
                if (frame_failed)
                    w.String(frames[index].filename().string().c_str());
                else
                    w.Null();
                demo::String(w, "message", error.what());
                w.EndObject();
                try
                {
                    report->Publish(false, buffer.GetString());
                }
                catch (const std::exception& publication)
                {
                    GetLogger().error("Report publication failed: ", publication.what(),
                                      "; retained report/spool under ",
                                      arguments.output_path.string());
                }
            }
            throw;
        }
        return 0;
    }
    catch (const TCLAP::ExitException& exit_request)
    {
        return exit_request.getExitStatus();
    }
    catch (const TCLAP::ArgException& error)
    {
        GetLogger().error("Invalid command line: ", error.error(), " for ", error.argId());
        return 1;
    }
    catch (const std::invalid_argument& error)
    {
        GetLogger().error("Invalid input: ", error.what());
        return 1;
    }
    catch (const std::exception& error)
    {
        GetLogger().error("Centroiding failed: ", error.what());
        return 2;
    }
}

File: examples/centroiding_models/run_centroiding.py
"""Run independent image frames through the generated centroiding Python wrapper."""

from __future__ import annotations

import argparse
import sys
import time
import image_sequence as images
import inference_output as reports
from centroiding_metadata import metadata
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path

import autoforge_deploy as ptaf
import numpy as np
from PIL import Image


@dataclass(frozen=True)
class DemoOptions:
    """Validated options for one wrapper-backed centroiding run."""

    model_path: Path
    image_path: Path
    target: str
    device_id: int
    output_path: Path | None = None
    overlays: bool = False


@dataclass(frozen=True)
class CentroidResult:
    """One normalized centroid mapped into model and source-image pixels."""

    normalized_x: float
    normalized_y: float
    model_input_x_px: float
    model_input_y_px: float
    original_x_px: float
    original_y_px: float


def parse_arguments(arguments: Sequence[str] | None = None) -> DemoOptions:
    """Parse and validate command-line arguments.

    Args:
        arguments: Optional arguments excluding the executable name.

    Returns:
        Validated demo options.
    """

    parser = argparse.ArgumentParser(
        description=(
            "Run an image or directory through a centroiding manifest or raw ONNX "
            "using the generated autoforge_deploy wrapper."
        )
    )
    parser.add_argument("model_path", type=Path)
    parser.add_argument("image_path", metavar="input_path", type=Path)
    parser.add_argument("--output", type=Path, dest="output_path")
    parser.add_argument("--overlays", action="store_true")
    parser.add_argument(
        "--target",
        choices=("manifest", "cpu", "cuda"),
        default="manifest",
        help="Use configured runtime settings or force one execution target.",
    )
    parser.add_argument("--device", type=int, default=0, dest="device_id")
    parsed = parser.parse_args(arguments)

    if parsed.overlays and parsed.output_path is None:
        parser.error("--overlays requires --output")
    if parsed.device_id < 0:
        parser.error("--device must be non-negative")

    return DemoOptions(
        model_path=parsed.model_path,
        image_path=parsed.image_path,
        target=parsed.target,
        device_id=parsed.device_id,
        output_path=parsed.output_path,
        overlays=parsed.overlays,
    )


def make_runtime(options: DemoOptions) -> ptaf.SRuntimeConfig:
    """Build a strict runtime override from validated options.

    Args:
        options: Validated target and device selection.

    Returns:
        Backend-neutral runtime configuration.
    """

    runtime = ptaf.SRuntimeConfig()
    runtime.SetDeviceId(options.device_id)
    runtime.ClearExecutionTargetPriority()
    target = ptaf.EExecutionTarget.cpu if options.target == "cpu" else ptaf.EExecutionTarget.cuda
    runtime.AddExecutionTarget(target)
    runtime.SetAllowFallback(False)
    return runtime


def load_model(options: DemoOptions) -> ptaf.CModelFacade:
    """Load a centroiding manifest or ONNX artifact.

    Args:
        options: Validated model and runtime selection.

    Returns:
        Loaded role-level facade.

    Raises:
        ValueError: If the artifact extension is unsupported.
        RuntimeError: If the loaded model contract is not image-only centroiding.
    """

    model = ptaf.CModelFacade()
    extension = options.model_path.suffix.lower()
    has_override = options.target != "manifest"

    if extension == ".ptafmodel":
        if has_override:
            model.LoadModelConfigWithRuntimeConfig(str(options.model_path), make_runtime(options))
        else:
            model.LoadModelConfig(str(options.model_path))
    elif extension == ".onnx":
        if has_override:
            model.LoadModelWithRoleAndRuntimeConfig(
                str(options.model_path),
                ptaf.EModelRole.centroiding,
                make_runtime(options),
            )
        else:
            model.LoadModelWithRole(str(options.model_path), ptaf.EModelRole.centroiding)
    else:
        raise ValueError("Expected a .ptafmodel manifest or .onnx artifact")

    if model.GetRole() != "centroiding":
        raise RuntimeError("Centroiding requires role=centroiding")
    if model.GetNumInputs() != 1 or model.GetNumOutputs() != 1:
        raise RuntimeError("Centroiding requires exactly one input and one output")
    return model


def prepare_image(
    model: ptaf.CModelFacade, image_path: Path
) -> tuple[ptaf.SFloatTensor, tuple[int, int]]:
    """Decode, resize, and convert one image to a wrapper-owned NCHW tensor.

    Args:
        model: Loaded single-input centroiding facade.
        image_path: Image path understood by Pillow.

    Returns:
        Prepared tensor and original ``(width, height)``.

    Raises:
        RuntimeError: If the input tensor contract is unsupported.
    """

    input_info = model.GetInputInfo(0)
    shape = list(input_info.shape)
    if (
        input_info.dtype != "float32"
        or len(shape) != 4
        or shape[0] not in (-1, 1)
        or shape[1] != 1
        or shape[2] <= 0
        or shape[3] <= 0
    ):
        raise RuntimeError("Centroiding expects float32 input [N,1,H,W] with concrete H/W")

    height, width = int(shape[2]), int(shape[3])
    with Image.open(image_path) as source_image:
        grayscale_image = source_image.convert("L")
        original_size = grayscale_image.size
        resized_image = grayscale_image.resize((width, height), Image.Resampling.BILINEAR)
        hwc_values = np.asarray(resized_image, dtype=np.float32).reshape(-1)

    tensor = ptaf.MakeNchwFloatTensorFromHwcFloat(
        input_info.name,
        hwc_values,
        height,
        width,
        1,
        1.0 / 255.0,
        False,
    )
    return tensor, original_size


def decode_centroid(
    output: ptaf.SFloatTensor,
    model_input_size: tuple[int, int],
    original_size: tuple[int, int],
) -> CentroidResult:
    """Decode and map one normalized wrapper output.

    Args:
        output: Sole model output with concrete shape ``[1,2]``.
        model_input_size: Model ``(width, height)``.
        original_size: Decoded image ``(width, height)``.

    Returns:
        Validated coordinates in all supported spaces.

    Raises:
        RuntimeError: If the selected output contract is invalid.
    """

    if list(output.shape) != [1, 2]:
        raise RuntimeError("Centroiding expects output shape [1,2]")

    schema = ptaf.SFeatureRowSchema()
    features = ptaf.DecodeFeatureRows(output, schema)
    if len(features) != 1:
        raise RuntimeError("Centroiding must produce exactly one feature")

    normalized_x = float(features[0].position.x)
    normalized_y = float(features[0].position.y)
    if not np.isfinite(normalized_x) or not np.isfinite(normalized_y):
        raise RuntimeError("Centroiding coordinates must be finite")

    model_width, model_height = model_input_size
    original_width, original_height = original_size
    return CentroidResult(
        normalized_x=normalized_x,
        normalized_y=normalized_y,
        model_input_x_px=normalized_x * model_width,
        model_input_y_px=normalized_y * model_height,
        original_x_px=normalized_x * original_width,
        original_y_px=normalized_y * original_height,
    )


def print_result(
    model: ptaf.CModelFacade,
    input_tensor: ptaf.SFloatTensor,
    output_tensor: ptaf.SFloatTensor,
    original_size: tuple[int, int],
    result: CentroidResult,
) -> None:
    """Print stable fields shared with the native and MATLAB demos.

    Args:
        model: Loaded role-level facade.
        input_tensor: Prepared model input.
        output_tensor: Raw model output.
        original_size: Decoded image ``(width, height)``.
        result: Validated centroid coordinates.
    """

    print(f"role={model.GetRole()}")
    print(f"backend={model.GetBackendDetail()}")
    print(f"input_name={input_tensor.name}")
    print(f"input_shape={list(input_tensor.shape)}")
    print(f"image_size={list(original_size)}")
    print(f"output_name={output_tensor.name}")
    print(f"output_shape={list(output_tensor.shape)}")
    print(f"centroid.normalized_x={result.normalized_x:.9g}")
    print(f"centroid.normalized_y={result.normalized_y:.9g}")
    print(f"centroid.model_input_x_px={result.model_input_x_px:.9g}")
    print(f"centroid.model_input_y_px={result.model_input_y_px:.9g}")
    print(f"centroid.original_x_px={result.original_x_px:.9g}")
    print(f"centroid.original_y_px={result.original_y_px:.9g}")


def run_demo(options: DemoOptions) -> None:
    """Run one model load and independent frame inferences with optional reports.

    Args:
        options: Validated demo options.

    Raises:
        FileNotFoundError: If the model or image does not exist.
    """

    frames = images.select_frames(options.image_path)
    report = None
    if options.output_path is not None:
        reports.prepare_output(options.output_path, options.image_path)
        report = reports.Report(
            options.output_path,
            metadata(options.image_path, len(frames), requested_model=options.model_path),
        )
    index = None
    source = None
    stage = "model_load"
    complete = False
    try:
        model = load_model(options)
        if report is not None:
            report.metadata = metadata(options.image_path, len(frames), model, options.model_path)
        if options.overlays:
            if options.output_path is None:
                raise ValueError("--overlays requires --output")
            (options.output_path / "overlays").mkdir()
        for index, source in enumerate(frames):
            stage = "preprocessing"
            input_tensor, original_size = prepare_image(model, source)
            stage = "inference"
            start = time.perf_counter()
            output_tensor = model.InferSingleFloatTensor(input_tensor)
            duration = (time.perf_counter() - start) * 1000.0
            stage = "decoding_output"
            model_size = (int(input_tensor.shape[3]), int(input_tensor.shape[2]))
            result = decode_centroid(output_tensor, model_size, original_size)
            print(f"frame.index={index}\nframe.source={source.name}\ninference_ms={duration}")
            print_result(model, input_tensor, output_tensor, original_size, result)
            overlay = None
            if options.overlays and options.output_path is not None:
                stage = "overlay"
                overlay = f"overlays/{index:06d}_{source.stem}.png"
                images.save_overlay(
                    source,
                    options.output_path / overlay,
                    result.original_x_px,
                    result.original_y_px,
                )
            stage = "report"
            if report is not None:
                report.append(
                    {
                        "index": index,
                        "source": source.name,
                        "image_size": {"width": original_size[0], "height": original_size[1]},
                        "raw_output": {
                            "name": output_tensor.name,
                            "shape": list(output_tensor.shape),
                            "values": [float(v) for v in output_tensor.values],
                        },
                        "centroid": {
                            "normalized": {"x": result.normalized_x, "y": result.normalized_y},
                            "model_pixels": {
                                "x": result.model_input_x_px,
                                "y": result.model_input_y_px,
                            },
                            "image_pixels": {"x": result.original_x_px, "y": result.original_y_px},
                            "inside_image": (
                                0 <= result.original_x_px < original_size[0]
                                and 0 <= result.original_y_px < original_size[1]
                            ),
                        },
                        "inference_ms": duration,
                        "overlay": overlay,
                    }
                )
        index = None
        source = None
        stage = "report"
        if report is not None:
            report.publish(True)
        complete = True
    except Exception as error:
        if report is not None:
            try:
                report.publish(
                    False,
                    {
                        "stage": stage,
                        "frame_index": index,
                        "source": source.name if source else None,
                        "message": str(error),
                    },
                )
            except OSError as publication:
                print(
                    f"Report publication failed: {publication}; retained files in {report.root}",
                    file=sys.stderr,
                )
        raise
    finally:
        if report is not None:
            report.close(complete)


def main(arguments: Sequence[str] | None = None) -> int:
    """Run the command-line demo and translate failures to process status.

    Args:
        arguments: Optional arguments excluding the executable name.

    Returns:
        Zero on success and one on a model, image, or runtime failure.
    """

    try:
        run_demo(parse_arguments(arguments))
    except (OSError, RuntimeError, ValueError, TypeError) as error:
        print(f"run_centroiding.py failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
